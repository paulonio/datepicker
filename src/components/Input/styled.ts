import styled from 'styled-components';

export const Wrapper = styled.div`
  width: 250px;
`;

export const Label = styled.div`
  margin-bottom: 8px;
  font-size: 15px;
  line-height: 20px;
  font-weight: 600;
  color: #333333;
`;

export const FieldWrapper = styled.label`
  position: relative;
  margin-bottom: 8px;
  display: block;
  border: solid 1px #dddddd;
  border-radius: 8px;
`;

export const LeftIconWrapper = styled.div`
  position: absolute;
  width: 16px;
  height: 16px;
  top: 13px;
  left: 15px;
  cursor: pointer;
`;

export const RightIconWrapper = styled.div`
  position: absolute;
  width: 16px;
  height: 16px;
  top: 14px;
  right: 15px;
  cursor: pointer;
`;

export const Field = styled.input`
  width: calc(100% - 78px);
  border: none;
  border-radius: 8px;
  padding: 11px 39px 11px 39px;
  font-size: 15px;
  line-height: 20px;
  font-weight: 400;
  background-color: #ffffff;
  color: #333333;
  &:focus {
    outline: none;
  }
`;
