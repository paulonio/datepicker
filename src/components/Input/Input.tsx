import React, { ChangeEvent, FC, useEffect, useState } from 'react';
import { CalendarIcon, CloseIcon, Field, FieldWrapper, Icon, Label, Wrapper } from './styled';

interface InputProps {
  label: 'From' | 'To';
  date: Date | null;
  onDateChange: (date: Date | null) => void;
  setCurrentCalendar: (value: 'From' | 'To') => void;
  // TODO fix
  toggleCalendar: (bool: boolean) => void;
}

export const updateInput = (date: Date | null) => {
  if (!date) {
    return '';
  }

  const year = date.getFullYear();
  const month = date.getMonth();
  const day = date.getDate();

  return `${day}/${month + 1}/${year}`;
};

const Input: FC<InputProps> = ({
  label,
  date,
  onDateChange,
  setCurrentCalendar,
  toggleCalendar,
}) => {
  const [inputDate, setInputDate] = useState<string>('');

  useEffect(() => {
    const dateToUpdate = updateInput(date);
    setInputDate(dateToUpdate);
  }, [date]);

  const updateSelectedDate = () => {
    if (!onDateChange) {
      return;
    }

    if (inputDate === '') {
      return;
    }

    const [day, month, year] = inputDate
      .trim()
      .split('/')
      .map((item) => Number(item));

    onDateChange(new Date(year, month - 1, day));
  };

  const handleFocus = () => {
    toggleCalendar(true);
    setCurrentCalendar(label);
  };

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    const value = e.target.value;
    setInputDate(value);
  };

  const resetInput = () => {
    setInputDate('');
    onDateChange(null);
  };

  return (
    <Wrapper>
      <Label>{label}</Label>
      <FieldWrapper>
        <CalendarIcon>
          <Icon src="./assets/icons/Calendar.svg" />
        </CalendarIcon>
        <Field
          value={inputDate}
          onChange={handleChange}
          onBlur={updateSelectedDate}
          onFocus={handleFocus}
          placeholder="Choose Date"
        />
        {inputDate && (
          <CloseIcon onClick={resetInput}>
            <Icon src="./assets/icons/Clear.svg" />
          </CloseIcon>
        )}
      </FieldWrapper>
    </Wrapper>
  );
};

export default Input;
