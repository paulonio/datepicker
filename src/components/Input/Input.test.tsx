import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import Input from './Input';

describe('Input', () => {
  const dateMock = new Date(2023, 4, 7);

  it('should render input', () => {
    const { container } = render(<Input label="From" date={dateMock} dispatch={() => {}} />);
    expect(container).toMatchSnapshot();
  });

  it('should render input with date', () => {
    render(<Input label="From" date={dateMock} dispatch={() => {}} />);
    const input: HTMLInputElement = screen.getByTestId('input');
    expect(input.value).toBe('07/05/2023');
  });

  it('should render empty input', () => {
    render(<Input label="From" date={null} dispatch={() => {}} />);
    const input: HTMLInputElement = screen.getByTestId('input');
    expect(input.value).toBe('');
  });

  it('should clear input', () => {
    render(<Input label="From" date={dateMock} dispatch={() => {}} />);
    const input: HTMLInputElement = screen.getByTestId('input');
    const closeButton = screen.getByTestId('close-icon');
    fireEvent.click(closeButton);
    expect(input.value).toBe('');
  });

  it('should correctly change date in the input', () => {
    render(<Input label="From" date={null} dispatch={() => {}} />);
    const input: HTMLInputElement = screen.getByTestId('input');
    fireEvent.change(input, { target: { value: '07/05/2023' } });
    fireEvent.blur(input);
    expect(input.value).toBe('07/05/2023');
  });

  it('should not save date on invalid value', () => {
    render(<Input label="From" date={null} dispatch={() => {}} />);
    const input: HTMLInputElement = screen.getByTestId('input');
    fireEvent.change(input, { target: { value: '01/01/1900' } });
    fireEvent.blur(input);
    expect(input.value).toBe('');
  });
});
