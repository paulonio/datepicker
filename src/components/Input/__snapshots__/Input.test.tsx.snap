// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render input 1`] = `
<div>
  <div
    class="sc-beqWaB eiKLBH"
  >
    <div
      class="sc-gueYoa kEZcTP"
    >
      From
    </div>
    <label
      class="sc-dmqHEX gTVuuB"
    >
      <div
        class="sc-hLseeU hJBuPB"
      >
        <img
          class="sc-jTrPJq cfnNIk"
          src="./assets/icons/Calendar.svg"
        />
      </div>
      <input
        class="sc-gLDzan bZWVsH"
        data-testid="input"
        placeholder="Choose Date"
        value="07/05/2023"
      />
      <div
        class="sc-eDDNvR eeFhds"
        data-testid="close-icon"
      >
        <img
          class="sc-jTrPJq cfnNIk"
          src="./assets/icons/Clear.svg"
        />
      </div>
    </label>
  </div>
</div>
`;
